[codemod] Add multiple codemods to migrate components from v4 to v5 (#24867)
mui-org/material-ui: docs/src/pages/guides/migration-v4/migration-v4.md

- Renamed `fade` to `alpha` to better describe its functionality.
  The previous name was leading to confusion when the input color already had an alpha value. The helper **overrides** the alpha value of the color.

  ```diff
  - import { fade } from '@material-ui/core/styles';
  + import { alpha } from '@material-ui/core/styles';

  const classes = makeStyles(theme => ({
  -  backgroundColor: fade(theme.palette.primary.main, theme.palette.action.selectedOpacity),
  +  backgroundColor: alpha(theme.palette.primary.main, theme.palette.action.selectedOpacity),
  }));
  ```

  You can use the [`fade-rename-alpha` codemod](https://github.com/mui-org/material-ui/tree/HEAD/packages/material-ui-codemod#fade-rename-alpha) for automatic migration.

> -The `createStyles` function from `@material-ui/core/styles` was moved to the one exported from `@material-ui/styles`. It is necessary for removing the dependency to `@material-ui/styles` in the core package.

Reviewer: ```suggestion
- The `createStyles` function from `@material-ui/core/styles` was moved to the one exported from `@material-ui/styles`. It is necessary for removing the dependency to `@material-ui/styles` in the core package.
```

+<Box sx={{ color: 'primary.main' }} />
```

> Note that the system grid function wasn't documented in v4.

### Core components

As the core components use emotion as their style engine, the props used by emotion are not intercepted. The prop `as` in the following code snippet will not be propagated to `SomeOtherComponent`.

> `<MuiComponent component={SomeOtherComponent} as="button" />`

Reviewer: Looks cleaner:

````suggestion
```jsx
<MuiComponent component={SomeOtherComponent} as="button" />`
````

### AppBar

> - [AppBar] Remove z-index when position static and relative (WUT?)

Reviewer: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context

```suggestion
- [AppBar] Remove z-index when position static and relative. This avoids the creation of a stacking context and rendering issues.
```

  +<Collapse classes={{ root: 'collapse' }}>
  ```

You can use the [`collapse-rename-collapsedheight` codemod](https://github.com/mui-org/material-ui/tree/HEAD/packages/material-ui-codemod#collapse-rename-collapsedheight) for automatic migration of both the prop and the classes key.

> ### CssBaseline

Reviewer: ```suggestion
### CssBaseline
```
